// scraperapi-sdk/dist/ScraperAPIClient.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fetch_1 = __importDefault(require("./util/fetch"));
/**
 * A backward compatible ScraperAPI client that provides a few shorthand methods to scrape websites easily.
 */
var ScraperAPIClient = /** @class */ (function () {
    function ScraperAPIClient(apiKey) {
        this.apiKey = apiKey;
    }
    /**
     * Function to retrieve account information.
     *
     * @returns {{
     *  "burst": number; // The number of burst (frequent) requests currently in flight
     *  "concurrencyLimit": number; // The maximum number of concurrent requests allowed
     *  "concurrentRequests": number; // The number of concurrent requests currently in flight
     *  "failedRequestCount": number; // The number of failed requests in the current billing cycle
     *  "requestCount": number; // The total number of requests in the current billing cycle
     *  "requestLimit": number; // The maximum number of requests allowed in the current billing cycle
     *  "subscriptionDate": Date; // The start date of the current billing cycle
     * }}
     */
    ScraperAPIClient.prototype.account = function () {
        return __awaiter(this, void 0, void 0, function () {
            var accountResponse, parsedAccountResponse;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, (0, fetch_1.default)('account', { api_key: this.apiKey }, 'GET')];
                    case 1:
                        accountResponse = (_a.sent()).body;
                        parsedAccountResponse = JSON.parse(accountResponse);
                        return [2 /*return*/, __assign(__assign({}, parsedAccountResponse), { subscriptionDate: new Date(parsedAccountResponse.subscriptionDate) })];
                }
            });
        });
    };
    /**
     * Shorthand function to send a GET request to the API and fetch the provided URL.
     *
     * @param url The target URL to fetch though the API, e.g. https://example.com
     * @param options The API options to use for this request (e.g. render, country_code, etc.)
     * @param headers The headers to be passed to the API (optional)
     * @returns APIResponse
     */
    ScraperAPIClient.prototype.get = function (url, options, headers) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, fetch_1.default)('/', __assign(__assign({}, options), { api_key: this.apiKey, url: url }), 'GET', headers)];
            });
        });
    };
    /**
     * Shorthand function to send a POST request to the API and fetch the provided URL.
     *
     * @param url The target URL to fetch though the API, e.g. https://example.com
     * @param options The API options to use for this request (e.g. render, country_code, etc.)
     * @param headers The headers to be passed to the API (optional)
     * @param body The body to be passed to the API (optional)
     * @returns APIResponse
     */
    ScraperAPIClient.prototype.post = function (url, options, headers, body) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, fetch_1.default)('/', __assign(__assign({}, options), { api_key: this.apiKey, url: url }), 'POST', headers, body)];
            });
        });
    };
    /**
     * Shorthand function to send a PUT request to the API and fetch the provided URL.
     *
     * @param url The target URL to fetch though the API, e.g. https://example.com
     * @param options The API options to use for this request (e.g. render, country_code, etc.)
     * @param headers The headers to be passed to the API (optional)
     * @param body The body to be passed to the API (optional)
     * @returns APIResponse
     */
    ScraperAPIClient.prototype.put = function (url, options, headers, body) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, (0, fetch_1.default)('/', __assign(__assign({}, options), { api_key: this.apiKey, url: url }), 'PUT', headers, body)];
            });
        });
    };
    return ScraperAPIClient;
}());
exports.default = ScraperAPIClient;
